misk-hibernate-testing/src/main/kotlin/misk/jdbc/AbstractScaleSafetyChecks.kt

```diff
@@ -0,0 +1,154 @@
+  val isScatter: Boolean
+    get() {
+      return Opcode == "SelectScatter" || Input?.isScatter ?: false
+    }
+}
+
+internal data class QueryPlan(
+  val Original: String,
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+internal data class Variables(
+  /** Queries processed by plan type */
+  val QueriesProcessed: Map<String, Int>
+)
+
+abstract class AbstractScaleSafetyChecks @Inject constructor(
+  open val config: DataSourceConfig,
+  open val okHttpClient: OkHttpClient,
+  open val moshi: Moshi,
+  open val transacter: Transacter
+) : DataSourceDecorator {
+  internal abstract fun connect(): Connection?
+
+  internal fun isDml(query: String): Boolean {
+    val first = query
+      .replace(COMMENT_PATTERN, "")
+      .trimStart()
+      .toLowerCase()
+      .takeWhile { !it.isWhitespace() }
+    return DML.contains(first)
+  }
+
+  /**
+   * Digs into the MySQL log to find the last executed DML statement that passed through Vitess.
+   */
+  internal fun extractLastDmlQuery(): String? {
+    return connect()?.let { c ->
+      c.createStatement().use { s ->
+        s.executeQuery("""
+                  SELECT argument
+                  FROM mysql.general_log
+                  WHERE command_type = 'Query'
+                  AND (
+                    argument LIKE '%update%'
+                    OR argument LIKE '%insert%'
+                    OR argument LIKE '%delete%'
+                  )
+                  AND NOT argument LIKE 'SELECT argument%'
+                  ORDER BY event_time DESC
+                  LIMIT 1
+                """.trimIndent())
+          .uniqueString()
+      }
+    }
+  }
+
+  /**
+   * Digs into the MySQL log to find the last executed DML statement that passed through Vitess.
+   */
+  internal open fun extractQueriesSince(mysqlTime: Timestamp): List<String> {
+    return connect()?.let { c ->
+      c.prepareStatement("""
+                  SELECT argument
+                  FROM mysql.general_log
+                  WHERE command_type in ('Query', 'Execute')
+                  AND NOT argument LIKE '%general_log%'
+                  AND NOT argument = 'begin'
+                  AND NOT argument LIKE '%1 != 1%'
+                  AND NOT lower(argument) LIKE '%information_schema%'
+                  AND NOT lower(argument) LIKE 'use %'
+                  AND NOT lower(argument) LIKE 'show %'
+                  AND NOT lower(argument) LIKE 'describe %'
+                  AND event_time > ?
+                  ORDER BY event_time DESC
+                """.trimIndent()).use { s ->
+        s.setTimestamp(1, mysqlTime)
+        s.executeQuery().map { it.getString(1) }
+      }
+    } ?: arrayListOf()
+  }
+
+  /**
+   * Figure out how many total full scatter queries we've executed so far.
+   */
+  internal fun extractScatterQueryCount(): Int {
+    val request = Request.Builder()
```

Comment:
should this be in a subtype?

Reply:
i'm not too sure what you mean. any chance you could show me an example?